configuration/src/main/java/io/airlift/configuration/ConfigurationInterpolator.java

```diff
@@ -0,0 +1,114 @@
+package io.airlift.configuration;
+
+import com.google.common.collect.ImmutableMap;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static com.google.common.base.Verify.verify;
+import static java.util.Locale.ENGLISH;
+import static java.util.Objects.requireNonNull;
+import static java.util.regex.Matcher.quoteReplacement;
+
+public class ConfigurationInterpolator
+{
+    private static final Pattern PATTERN = Pattern.compile("\\$\\{(?<name>[A-Z]+):(?<key>[a-zA-Z0-9_\\-/]*)}");
+    private final Map<String, ValueInterpolator> valueProviders;
+
+    public ConfigurationInterpolator(Map<String, ValueInterpolator> valueProviders)
+    {
+        this.valueProviders = ImmutableMap.copyOf(requireNonNull(valueProviders, "valueProviders is null"));
+        verifyValueProviderNames(valueProviders.keySet());
+    }
+
+    private static void verifyValueProviderNames(Set<String> names)
+    {
+        for (String name : names) {
+            verify(name.toUpperCase(ENGLISH).contentEquals(name), "ValueInterpolator name '%s' is not uppercase", name);
+        }
+    }
+
+    public Map<String, String> interpolate(Map<String, String> properties, BiConsumer<String, String> onError)
+    {
+        Map<String, String> replaced = new HashMap<>();
+        properties.forEach((propertyKey, propertyValue) -> {
+            StringBuilder replacedPropertyValue = new StringBuilder();
+            Matcher matcher = PATTERN.matcher(propertyValue);
+            while (matcher.find()) {
+                String interpolatorName = matcher.group("name");
+                String keyName = matcher.group("key");
+                ValueInterpolator valueInterpolator = valueProviders.get(interpolatorName);
+                if (valueInterpolator == null) {
+                    onError.accept(propertyKey, "Configuration property '%s' references value interpolator '%s' that does not exist".formatted(propertyKey, interpolatorName));
+                    return;
+                }
+
+                Optional<String> replacedValue = valueInterpolator.interpolate(keyName);
+                if (replacedValue.orElse(null) == null) {
```

Comment:
Would `isEmpty()` be more idiomatic? Should we assert that `replacedValue` is not null? That would signal some other issues with the interpolator, right?
```suggestion
                if (replacedValue.isEmpty()) {
```